src/skadi/sample.cc

```diff
@@ -449,11 +553,79 @@ template <class coords_t> std::vector<double> sample::get_all(const coords_t& co
       value = tile.get(u, v);
     }
 
+    if (value == get_no_data_value()) {
+      if (!seen.count(coord)) {
+        coord_st.push_back(coord);
+        seen.insert(coord);
+      }
+      continue;
+    }
+
     values.emplace_back(value);
   }
+
+  std::function<double(const decltype(elem)&)> func = [this](const decltype(elem)& param) -> double {
+    return this->get_from_remote(param);
+  };
+
+  parallel_call(func, coord_st, values, num_threads_);
+
   return values;
 }
 
+template <class coord_t> double sample::get_from_cache(const coord_t& coord) {
+  auto tile = cache_->source(get_tile_index(coord));
+  if (!tile)
+    return get_no_data_value();
+
+  double u = (coord.first - std::floor(coord.first)) * (HGT_DIM - 1);
+  double v = (1.0 - (coord.second - std::floor(coord.second))) * (HGT_DIM - 1);
+  return tile.get(u, v);
+}
+
+template <class coord_t> double sample::get_from_remote(const coord_t& coord) {
+  if (url_.empty() || !remote_loader_)
+    return get_no_data_value();
+
+  auto index = get_tile_index(coord);
+  auto elev = get_hgt_file_name(index);
+  auto uri = make_single_point_url(url_, elev);
+
+  {
+    std::lock_guard<std::mutex> _(st_lck_);
+    if (st_.count(uri))
+      return get_from_cache(coord);
+
+    st_.insert(uri);
+  }
+
+  LOG_INFO("Start loading data from remote server address: " + uri);
+  int repeat{3};
+  auto result = remote_loader_->get(uri);
+  while (repeat-- > 0 && result.status_ != baldr::tile_getter_t::status_code_t::SUCCESS) {
+    result = remote_loader_->get(uri);
```

Comment:
Perhaps do some sleep here with exponential backoff?

Reply:
at the end decided to remove retry from here, it seems that its place in cur.h at the end. but left this improvement for later